includes/class-wc-stripe-api.php

```diff
@@ -176,4 +176,97 @@ public static function retrieve( $api ) {
 
 		return json_decode( $response['body'] );
 	}
+
+	/**
+	 * Send the request to Stripe's API with level 3 data generated
+	 * from the order. If the request fails due to an error related
+	 * to level3 data, make the request again without it to allow
+	 * the payment to go through.
+	 *
+	 * @since 4.3.2
+	 * @version 4.3.2
+	 *
+	 * @param array    $request     Array with request parameters.
+	 * @param string   $api         The API path for the request.
+	 * @param array    $level3_data The level 3 data for this request.
+	 * @param WC_Order $order       The order associated with the payment.
+	 *
+	 * @return stdClass|array The response
+	 */
+	public static function request_with_level3_data( $request, $api, $level3_data, $order ) {
+		// Do not add level3 data it's the array is empty.
+		if ( empty( $level3_data ) ) {
+			return self::request(
+				$request,
+				$api
+			);
+		}
+
+		// If there's a transient indicating that level3 data was not accepted by
+		// Stripe in the past for this account, do not try to add level3 data.
+		if ( get_transient( 'wc_stripe_level3_not_allowed' ) ) {
+			return self::request(
+				$request,
+				$api
+			);
+		}
+
+		// Add level 3 data to the request.
+		$request['level3'] = $level3_data;
+
+		$result = self::request(
+			$request,
+			$api
+		);
+
+		$is_level3_param_not_allowed = (
+			isset( $result->error )
+			&& isset( $result->error->code )
+			&& 'parameter_unknown' === $result->error->code
+			&& isset( $result->error->param )
+			&& 'level3' === $result->error->param
+		);
+
+		if ( $is_level3_param_not_allowed ) {
+			// Set a transient so that future requests do not add level 3 data.
+			// Transient is set to expire in 3 months, can be manually removed if needed.
+			set_transient( 'wc_stripe_level3_not_allowed', true, 3 * MONTH_IN_SECONDS );
+
+			// Make the request again without level 3 data.
+			unset( $request['level3'] );
+			$result = WC_Stripe_API::request(
+				$request,
+				$api
+			);
+		}
```

Comment:
I think it would be better to just `return` instead of setting the value of `$result` since AFAICT you can't enter the next `if` statement anyway.

Reply:
Why not? The purpose of the `if` statement is to log `invalid_request_error` requests in order to be able to debug them. If the response from the API contains such an error, this can (and does) happen.

Reply:
My expectation was that `is_level_3data_incorrect` (the check before the next `if` statement) would never be true because `$request['level3']` is unset before sending the data in **this** `if` statement, i.e. how can the level 3 data be incorrect if the level 3 data wasn't sent (because level 3 data isn't supported).

If we can still enter the second `if` statement after having sent a second request, I'd suggest naming the check something other than `$is_level_3data_incorrect`, e.g. just `$is_request_data_incorrect` or something along those lines.

In essence I think the variable name of `$is_level_3data_incorrect` caused some confusion here for me 🙂.

Reply:
Although I did not rename the variable, I see how it could be confusing, so I restructured it here: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1130/commits/880fddc08e0a5cb56d2aad74b460b2968fae56c0. What do you think?

Reply:
Perfect 👌